compressors/rANS.py

```diff
@@ -0,0 +1,335 @@
+            )
+            out_bits = _bits + out_bits
+            state = state >> self.params.NUM_BITS_OUT
+
+        return state, out_bits
+
+    def encode_symbol(self, s, state: int) -> Tuple[int, BitArray]:
+        """Encodes the next symbol, returns some bits and  the updated state
+
+        Args:
+            s (Any): next symbol to be encoded
+            state (int): the rANS state
+
+        Returns:
+            state (int), symbol_bitarray (BitArray):
+        """
+        # output bits to the stream so that the state is in the acceptable range
+        # [L, H] *after*the `rans_base_encode_step`
+        symbol_bitarray = BitArray("")
+        state, out_bits = self.shrink_state(state, s)
+        symbol_bitarray = out_bits + symbol_bitarray
+
+        # core encoding step
+        state = self.rans_base_encode_step(s, state)
+        return state, symbol_bitarray
+
+    def encode_block(self, data_block: DataBlock):
+        # initialize the output
+        encoded_bitarray = BitArray("")
+
+        # initialize the state
+        state = self.params.initial_state(self.freqs)
+
+        # update the state
+        for s in data_block.data_list:
+            state, symbol_bitarray = self.encode_symbol(s, state)
+            encoded_bitarray = symbol_bitarray + encoded_bitarray
+
+        # Finally, pre-pend binary representation of the final state
+        num_state_bits = self.params.num_state_bits(self.freqs.total_freq)
+        encoded_bitarray = uint_to_bitarray(state, num_state_bits) + encoded_bitarray
+
+        # add the data_block size at the beginning
+        # NOTE: rANS decoding needs a way to indicate where to stop the decoding
+        # One way is to add a character at the end which signals EOF. This requires us to
+        # change the probabilities of the other symbols. Another way is to just signal the size of the
+        # block. These two approaches add a bit of overhead.. the approach we use is much more transparent
+        encoded_bitarray = (
+            uint_to_bitarray(data_block.size, self.params.DATA_BLOCK_SIZE_BITS) + encoded_bitarray
+        )
+
+        return encoded_bitarray
+
+
+class rANSDecoder(DataDecoder):
+    def __init__(self, freqs: Frequencies, rans_params: rANSParams):
+        self.freqs = freqs
+        self.params = rans_params
+
+    @staticmethod
+    def find_bin(cumulative_freqs_list: List, slot: int) -> int:
+        """Performs binary search over cumulative_freqs_list to locate which bin
+        the slot lies.
+
+        Args:
+            cumulative_freqs_list (List): the sorted list of cumulative frequencies
+                For example: freqs_list = [2,7,3], cumulative_freqs_list [0,2,9]
+            slot (int): the value to search in the sorted list
+
+        Returns:
+            bin: the bin in which the slot lies
+        """
+        bin = np.searchsorted(cumulative_freqs_list, slot, side="right") - 1
```

Comment:
add comment on why side="right" is important, maybe as part of example in doc comment

Reply:
done